components/normalizer/src/lib.rs

```diff
@@ -1608,5 +1621,262 @@ impl ComposingNormalizer {
     normalizer_methods!();
 }
 
+/// A struct for providing the raw canonical composition operation.
+pub struct CanonicalComposition {
+    canonical_compositions: DataPayload<CanonicalCompositionsV1Marker>,
+}
+
+impl CanonicalComposition {
+    /// Performs canonical composition (including Hangul) on a pair of
+    /// characters or returns `None` if these characters don't compose.
+    /// Composition exclusions are taken into account.
+    #[inline]
+    pub fn compose(&self, starter: char, second: char) -> Option<char> {
+        let v = u32::from(second).wrapping_sub(HANGUL_V_BASE);
+        if v >= HANGUL_T_LIMIT - HANGUL_V_BASE {
+            return self.compose_non_hangul(starter, second);
+        }
+        if v < HANGUL_V_COUNT {
+            let l = u32::from(starter).wrapping_sub(HANGUL_L_BASE);
+            if l < HANGUL_L_COUNT {
+                let lv = l * HANGUL_N_COUNT + v * HANGUL_T_COUNT;
+                // Safe, because the inputs are known to be in range.
+                return Some(unsafe { char::from_u32_unchecked(HANGUL_S_BASE + lv) });
+            }
+            return None;
+        }
+        if in_inclusive_range(second, '\u{11A8}', '\u{11C2}') {
+            let lv = u32::from(starter).wrapping_sub(HANGUL_S_BASE);
+            if lv < HANGUL_S_COUNT && lv % HANGUL_T_COUNT == 0 {
+                let lvt = lv + (u32::from(second) - HANGUL_T_BASE);
+                // Safe, because the inputs are known to be in range.
+                return Some(unsafe { char::from_u32_unchecked(HANGUL_S_BASE + lvt) });
+            }
+        }
+        None
+    }
+
+    /// Performs non-Hangul canonical composition on a pair of characters
+    /// or returns `None` if these characters don't compose. Composition
+    /// exclusions are taken into account. (The return value is `None` with
+    /// Hangul arguments.)
+    #[inline(always)]
+    fn compose_non_hangul(&self, starter: char, second: char) -> Option<char> {
```

Comment:
Question: why is `hangul` special-cased?

Reply:
Hangul is special-cased, because Hangul handling is algorithmic, so a caller might hard-code it and only call to an external library for the operation that needs data.

However, maybe having distinct API entry points to cater to this as a micro-optimization is the wrong call, since the Hangul check adds just one branch. I'll make the non-Hangul case private.

Reply:
Looks like non-Hangul was non-`pub` already.